pkg/runtime/conversion_generator.go

```diff
@@ -302,11 +302,15 @@ func (g *conversionGenerator) generateConversionsForSlice(inType, outType reflec
 }
 
 func (g *conversionGenerator) generateConversionsForStruct(inType, outType reflect.Type) error {
+	errs := []string{}
```

Comment:
reproducer added.  As @liggitt says, the generation for the parent type will return an error but the generation of the symmetrical subtypes will not and will produce the public, fast conversion method instead of falling back to reflection for all subtypes.  Then the custom conversion can make use of the public subtype conversions